src/v/datalake/record_multiplexer.h

```diff
@@ -55,11 +71,36 @@ class record_multiplexer {
       table_creator&,
       model::iceberg_invalid_record_action,
       location_provider,
-      translation_probe&,
-      lazy_abort_source& as);
+      translation_probe&);
 
-    ss::future<ss::stop_iteration> operator()(model::record_batch batch);
-    ss::future<result<write_result, writer_error>> end_of_stream();
+    /**
+     * Multiplex the data from a reader into writers per schema and partition.
+     * Can be called multiple times in succession before calling finish().
+     */
+    ss::future<> multiplex(
+      model::record_batch_reader reader,
+      model::timeout_clock::time_point deadline,
+      ss::abort_source& as);
+
+    /**
+     * Abortable multiplexing on a single batch. Visible for testing.
+     */
+    ss::future<ss::stop_iteration>
+    do_multiplex(model::record_batch batch, ss::abort_source&);
+
+    /**
+     * Forces a flush on all the underlying file writers resulting in freeing
+     * up buffered in flight writes. May not be called in parallel with other
+     * methods.
+     */
+    ss::future<writer_error> flush_writers();
```

Comment:
Should probably be noted in a comment that finish() must be called even in case of errors in flush. Might not be obvious but otherwise we are left with dangling writers/ostreams like in https://github.com/redpanda-data/redpanda/pull/25215/files